### DISCUSSION TITLE
SALT2 fit doesn't match datapoints

### DISCUSSION THREAD

**Discussion by skressie** [sncosmo/sncosmo#342] [Q&A]
Hi there! 
I struggle for a long while to create a suitable fit to my data and since I'm running out of ideas, time and hope, I would be so happy if someone could help me out and have a look over my code. 

I used data from a published supernova fitted with SALT2 so I can compare my results. 
I converted the magnitudes via 
`vega.band_mag_to_flux(magnitude, band)`

This is my relevant code: 
```
vega.band_mag_to_flux(magnitude, band)

dust = sncosmo.CCM89Dust()
model = sncosmo.Model(source='salt2', 
                      effects=[dust, dust], 
                      effect_names=['host','mw'],
                      effect_frames=['rest','obs'])

model.set(z=0.000804)
model.set(mwebv=0.0088)
model.set_source_peakabsmag(-19.2, 'bessellb', 'vega') #effects x0 of model, not fitted model. Without: x0=1

result, fitted_model = sncosmo.fit_lc(data, model, ['t0', 'x0', 'x1', 'c'], phase_range=(-16, 25))

fig = sncosmo.plot_lc(data, model=fitted_model, errors=result.errors)
```

What I tried so far, among other things:

- adapt hostebv / r_v as a parameter  -->  RunTime Error and values like hostebv = -0.016629,  hostr_v = +63280.407505. 
- vary phase range --> no effect 
- vary zeropints (just for testing) --> affects result of peakmagnitude, but not the fit 
- set z as a free paramter
- guess_amplitude True / False

Further, I wonder a bit about the peak magnitude. 
From my understanding, the following lines should do excatly the same, but they don't: 

`fitted_model.source_peakmag('bessellb', magsys='vega')) `
and 
```
x0_fit=fitted_model.get('x0')
vega.band_flux_to_mag(x0_fit, 'bessellb')
```
And last: I have to work with collected published data, so mostly I don't have information about zeropoints and systems. 
I guess there is no way to overcome this issue?

Thanks in advance!
Cheers, Sabrina

![SN2001el_LCfit](https://user-images.githubusercontent.com/94974833/182382292-ce02196c-89a5-413a-af63-4a542c0a1c8c.jpg)

**Comment by kbarbary** [comment]
Hi Sabrina! Sorry you were having trouble and didn't get a reply here. I have a few comments for you:

First, it looks like the fit is as good as it can be given the data: that is, the fitter found the best parameters possible: while adjusting the relative amplitude in different bands could improve the fit, the SALT2 model can't fit arbitrary offsets in each bands (by design): the `c` parameter adjusts bluer bands relative to redder bands, so decreasing `c` could make the _b_-band and _i_-band fit better, but at the expense of making the _v_- and _r_-bands fit worse.

This tells me that either (1) this SN has a really weird spectrum that doesn't match the SALT2 model, or (2) you made an error with the zeropoints or bandpasses of the data.

Was this SN really observed through Bessell filters? The filters should be the actual ones the data was collected in (or adjusted to).

The zeropoints and zeropoint system of the data are important, and can cause the fit you're seeing if not interpreted correctly. (There should really be an entire page in the sncosmo documentation on this; it is a common problem.)

Based on the first line of your code, I'm guessing that you have a table of magnitudes and bands, and you're converting those to flux. Are you sure the magnitudes you have are really in the Vega system? If they're actually in the AB system, you might see something like the fit you're seeing.

Some background: The "zeropoints" in sncosmo photometric data are somewhat arbitrary: you can multiply the flux by a factor of 10, and increase the zeropoint by 2.5, and the implied brightness will be the same (because `magnitude = zeropoint - 2.5 * log10(flux)` and magnitude is the thing that matters). In your code, if you do `vega.band_mag_to_flux(magnitude, band)`, that gives the flux corresponding to a zeropoint of `0.0`.

The magnitude system is the thing the magnitude is relative to: a magnitude of 20.0 in Vega means that the SN is a factor of `10**(20/2.5)` dimmer than the Vega spectrum in that particular band. A magnitude of 20.0 in AB means some thing different: the SN is a factor of `10**(20/2.5)` dimmer than the AB spectrum in that particular band. Changing the band or the magnitude system changes the meaning of the "magnitude" in terms of exactly how bright the SN is. And because Vega and AB have different relative brightnesses in different bands, assigning the wrong magnitude system to the data can cause relative offsets between bands.

**Comment by kbarbary** [comment]
A few other notes:

- You shouldn't need to add a "host dust" effect when using the SALT2 model. The `c` parameter of the model is designed to capture host dust.

- `x0` is an abritrary model scaling parameter, similar to "amplitude" in other models, and has no intrinsic meaning in terms of flux through any particular band. So this code:
  ```python
  x0_fit=fitted_model.get('x0')
  vega.band_flux_to_mag(x0_fit, 'bessellb')
  ```
  doesn't do what you want.

- 
  > And last: I have to work with collected published data, so mostly I don't have information about zeropoints and systems. I guess there is no way to overcome this issue?

  If you have published magnitudes, the authors should also say what magnitude system they are reporting magnitudes in. (The magnitudes themselves are useless without this information!)

**Comment by skressie** [comment]
Hi Kyle, thank you so much for your answer! 

You are right, I have a table with magnitudes I convert to flux via the built-in function. Since I have more or less the same problem with other SNe, I assume the problem is not the light curve of this particular event. I updated the picture with another SN that is well published, so I can be sure that it's actually in vega system and based on bessell filter systems. So the only problematic part left could be the zeropoints, I guess. The SN was published with vega zeropoints. I converted the magnitudes with these zp's by via `f = 10**(-0.4mag)*zp ` to see if it's in agreement with the sncomso built-in function `vega.band_mag_to_flux(magnitude, band)`, which was the case. Thus I conclude the zp in the astropy table (given in magnitudes?) should be 0. Is this conclusion correct? 

Regarding the other comment:

> x0 is an abritrary model scaling parameter, similar to "amplitude" in other models, and has no intrinsic meaning in terms of flux through any particular band.

If x0 is not the value I can use for calculate mB, could you tell me what's behind the built-in-function `fitted_model.source_peakmag('bessellb', magsys='vega')` so I can try to track the mistakes?

**Reply by kbarbary** [reply]
I'm not sure about the zeropoint thing. Can you point to where you're getting the data from? That might help.

Regarding x0, some more explanation might help: x0 scales the entire spectral timeseries up and down. So doubling x0 (and keeping all other parameters constant will increase the flux through any given band by a factor of two. For example:
```python
mag = model.source_peakmag("bessellb", magsys="vega")  # get peak mag
model['x0'] = 2 * model['x0']  # scale x0 by a factor of 2
new_mag = model.source_peakmag("bessellb", magsys="vega")  # get peak mag again
# new_mag should be  = mag - 2.5 * log10(2)
```

However, x0 is not the only parameter that changes the spectral time series; 'c' and 'x1' do as well. Any change to the spectral time series changes the flux through a band, and can even change the time of peak (slightly). Because of this, there is not a direct relationship between x0 and peak mag though a particular bandpass; the magnitude depends on all the parameters, not just x0.

`model.source_peakmag` works by generating the lightcurve in the given bandpass (given all the model parameters), finding the phase of the peak, and returning the value at the peak phase. You can see the implementation [here](https://github.com/sncosmo/sncosmo/blob/9fb781e354ca6edbc9e6f3df31da6b397e077dce/sncosmo/models.py#L439-L443).

**Comment by skressie** [comment]
All right, got the point with x0, thanks! 

Regarding the data, for example: 

Photometry: https://iopscience.iop.org/article/10.1086/345571/pdf  (Table 3 (combined with Table 4 to match Bessell Filters))
ZP: They only refer to the typical Vega zeropoints from Bessell 1998: https://articles.adsabs.harvard.edu/pdf/1998A%26A...333..231B (table A2). These zp's were used for the conversion from magnitude to flux via  `vega.band_mag_to_flux(magnitude, band)`, therefore I used zp 0 in the table.

**Comment by kbarbary** [comment]
The use of zeropoints from the Bessell 1998 paper is not necessary. The published magnitudes in the Krisciunas et al paper (Tables 3 and 4) should be all you need.

To convert from those magnitudes to flux, choose an arbitrary zeropoint, and do the following:
```python
# suppose `mag` is an array of Vega magnitudes
# and `magerr` is an array of uncertainties in those magnitudes

zp = 15.0  # can be anything, and can be the same for all bands.
flux = 10**(0.4 * (zp - mag))
fluxerr = flux * (math.log(10) / 2.5) * magerr

data = Table(
    {
        "time": ...,
        "band": ...,
        "flux": flux,
        "fluxerr": fluxerr,
        "zp": zp,
        "magsys": "vega",
    }
)
```

**Reply by skressie** [reply]
Hi Kyle, 

with this conversion of the magnitudes it works perfectly! I wonder a bit why this is not the same as `vega.band_mag_to_flux(magnitude, band)`, but anyway, I just will use this formular.

Thank you so much for your time & help, you saved my (scientific) life :)!
